src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs

```diff
@@ -799,17 +799,15 @@ public NodeWarningProperties GetIntersect(NodeWarningProperties other)
                 }
                 else if (thisPackages != null && otherPackages != null)
                 {
-                    packages = new Dictionary<string, HashSet<NuGetLogCode>>(StringComparer.OrdinalIgnoreCase);
+                    var sizeUpperBound = Math.Max(thisPackages.Count, otherPackages.Count);
+                    packages = new Dictionary<string, HashSet<NuGetLogCode>>(sizeUpperBound, StringComparer.OrdinalIgnoreCase);
 
-                    var allKeys = thisPackages.Keys.Concat(otherPackages.Keys)
-                        .Distinct(StringComparer.OrdinalIgnoreCase);
-
-                    foreach (var key in allKeys)
+                    foreach (var kv in thisPackages)
```

Comment:
Thank you for your contribution. 
Could you be able to include after change flame graph? We need to measure if it's improving or not.
Also, I believe there is correctness concern too.
`thisPackage` -> `a, b,c` 
`thisPackage` -> ` b,c,d`
Before your change result it `a, b, c ,d`
After your change `a, b, c` only.
You need to add another for loop iteration for `otherPackages` too.